ml-agents-envs/mlagents/envs/subprocess_environment.py

```diff
@@ -27,15 +27,24 @@ class UnityEnvWorker(NamedTuple):
     conn: Connection
 
     def send(self, name: str, payload=None):
-        cmd = EnvironmentCommand(name, payload)
-        self.conn.send(cmd)
+        try:
+            cmd = EnvironmentCommand(name, payload)
+            self.conn.send(cmd)
+        except (BrokenPipeError, EOFError):
+            raise KeyboardInterrupt
 
     def recv(self) -> EnvironmentResponse:
-        response: EnvironmentResponse = self.conn.recv()
-        return response
+        try:
+            response: EnvironmentResponse = self.conn.recv()
+            return response
+        except (BrokenPipeError, EOFError):
+            raise KeyboardInterrupt
```

Comment:
Why do you raise KeyboardInterrupt here?

Reply:
This should really probably be some custom exception, but in our case KeyboardInterrupt is a proxy for training being interrupted.  This is actually being sent from the main process which has the trainer and SubprocessUnityEnvironment.  If the pipe is broken or we receive an EOF, it means the worker has died before we told it to close -- this means either a keyboard interrupt happened (like on Windows) and the worker died before the main process or the environment process was manually destroyed, etc.  This way we'll still save the model.